### ISSUE TITLE
PageBy does not work in some cases. Workaround for the previous EfCore version does not work either

### ISSUE THREAD

**Issue by alexandis** [dotnet/efcore#28415]
`query.PageBy(skipCount, maxResultCount).ToListAsync(cancellationToken);`

In some cases for some tables, paging does not work properly - I can't be 100% sure, but seems like it happens when doing sorting by the column which has the same value in various rows of the page and THEN - trying to go to a different page. In EfCore 5 it was fixed by making this kludge

`query.Take(int.MaxValue).PageBy(skipCount, maxResultCount).ToListAsync(cancellationToken);`

which added intermediate `SELECT` before paging.

Now it does not work anymore: `Take(int.MaxValue)` is translated "as is".

How to resolve the issue now? It if matters, DB is Oracle, but probably the issue can be reproduced in other DBs.

**Comment by ErikEJ** [comment]
Where is "PageBy" defined?

**Comment by alexandis** [comment]
> Where is "PageBy" defined?

It is a 3rd party library extension. We use "ABP Framework". I took a look at their source code, but it's nothing fancy, just:

`query.Skip(skipCount).Take(maxResultCount);`

under the hood.

**Comment by roji** [comment]
> In some cases for some tables, paging does not work properly - I can't be 100% sure, but seems like it happens when doing sorting by the column which has the same value in various rows of the page and THEN - trying to go to a different page. In EfCore 5 it was fixed by making this kludge

It sounds like your resultset isn't ordered in a deterministic way; if you have the same value multiple times in the column by which you're ordering, then two paging queries which fetch different pages could show duplicate rows or omit them entirely, as  rows are ordered different across the two executions. The solution is to add a second, unique order column, typically the ID. See our [pagination docs](https://docs.microsoft.com/en-us/ef/core/querying/pagination) for more information.

If this isn't your problem, then we're going to need a runnable repro, without any 3rd-party library, which reproduces the problem.

**Comment by alexandis** [comment]
@roji Thank you. Seems like this is our case. I am closing the ticket.

**Comment by alexandis** [comment]
One small question: I can't figure out, how this is feasible in my situation.

The matter is, we use a common approach: text representation of a column name + sorting order is sent from client to server, which is translated to the proper `linq` query.

When I just add my DB indexes via fluent API to model binding method, it does not help to resolve the issue. But I can't use 'per table' approach for each `GetListAsync` method, it is way too complicated, we have tons of entities - each with the own key field (composite or simple) and own indexes. Is there a common approach to modify modelbinding for each entity and be sure ordering works properly for each sorting field without need to modify each `GetListAsync` call?

**Comment by roji** [comment]
@alexandis are you asking for some common way to automatically add a unique ordering column for all queries to make the results deterministic? If so, that's not going to be easy. It's probably possible to insert a custom expression visitor into EF's query pipeline, but that's complicated and could be brittle. In addition, this sort of "behind the scenes" manipulation of the query tree makes things magical and implicit - you can't really know what's going on just by looking at your source code.

In other words, I'd really recommend going over your queries and fixing them.
